backend/routes.js

```diff
@@ -247,3 +248,115 @@ route.GET('/app/{path*}', {}, {
 route.GET('/', {}, function (req, h) {
   return h.redirect('/app/')
 })
+
+route.POST('/zkpp/{contract}', {
+  validate: {
+    payload: Joi.alternatives([
+      {
+        b: Joi.string().required()
+      },
+      {
+        r: Joi.string().required(),
+        s: Joi.string().required(),
+        sig: Joi.string().required(),
+        Eh: Joi.string().required()
+      }
+    ])
+  }
+}, async function (req, h) {
+  try {
+    if (req.payload['b']) {
+      const result = await registrationKey(req.params['contract'], req.payload['b'])
+
+      if (result) {
+        return result
+      }
+    } else {
+      const result = await register(req.params['contract'], req.payload['r'], req.payload['s'], req.payload['sig'], req.payload['Eh'])
+
+      if (result) {
+        return result
+      }
+    }
+  } catch (e) {
+    const ip = req.info.remoteAddress
+    console.error('Error at POST /zkpp/{contract}: ' + e.message, { ip })
+  }
+
+  return Boom.internal('internal error')
+})
+
+route.GET('/zkpp/{contract}/auth_hash', {}, async function (req, h) {
+  if (!req.query['b']) {
+    return Boom.badRequest('b query param required')
+  }
+
+  try {
+    const challenge = await getChallenge(req.params['contract'], req.query['b'])
+
+    if (challenge) {
+      return challenge
+    }
+  } catch (e) {
+    const ip = req.info.remoteAddress
+    console.error('Error at GET /zkpp/{contract}/auth_hash: ' + e.message, { ip })
+  }
+
+  return Boom.internal('internal error')
+})
+
+route.GET('/zkpp/{contract}/contract_hash', {}, async function (req, h) {
+  if (!req.query['r']) {
+    return Boom.badRequest('r query param required')
+  }
+
+  if (!req.query['s']) {
+    return Boom.badRequest('s query param required')
+  }
+
+  if (!req.query['sig']) {
+    return Boom.badRequest('sig query param required')
+  }
+
+  if (!req.query['hc']) {
+    return Boom.badRequest('hc query param required')
+  }
+
+  try {
+    const salt = await getContractSalt(req.params['contract'], req.query['r'], req.query['s'], req.query['sig'], req.query['hc'])
+
+    if (salt) {
+      return salt
+    }
+  } catch (e) {
+    const ip = req.info.remoteAddress
+    console.error('Error at GET /zkpp/{contract}/contract_hash: ' + e.message, { ip })
+  }
+
+  return Boom.internal('internal error')
+})
+
+route.PUT('/zkpp/{contract}', {
```

Comment:
What is the difference between `route.POST('/zkpp/{contract}'` and `route.PUT('/zkpp/{contract}'` ?

Could you come up with another name for `route.PUT('/zkpp/{contract}'` (that's clear) so that we can stick with `POST`?

Reply:
`POST` is for new registrations and `PUT` is for updates, as is per RFC2616 section 2.6. I'm unsure what could be more descriptive other than a suffix (e.g., `/register`).

Reply:
In the new PR there are in-line code comments for how to update these APIs to use `POST` and more descriptive APIs